Add support for Iris's new text sink API

SquidDev opened the pull request:
This is needed on Iris 1.2.5 as they change the vertex format of `RenderType.text`, so us chucking bytes in directly won't work!

I'd like to profile this before merging and check that the new virtual calls are JITted away, but otherwise this is good for review.

CCing @IMS212, in case they have any comments.

SquidDev committed:
Add support for Iris's new text sink API

toad-dev commented on src/main/java/dan200/computercraft/client/render/TileEntityMonitorRenderer.java:
```diff
@@ -191,28 +191,27 @@ private static void renderTerminal( @Nonnull MultiBufferSource bufferSource, Pos
+                    DirectFixedWidthFontRenderer.drawCursor( sink, 0, 0, terminal, !monitor.isColour() );
 
                     buffer.flip();
-                    vbo.upload( termVertices, RenderTypes.MONITOR.mode(), RenderTypes.MONITOR.format(), buffer );
+                    vbo.upload( buffer.position() / vertexSize, RenderTypes.MONITOR.mode(), sink.getFormat(), buffer );
```
`buffer.flip()` should be moved to after the `vbo.upload( buffer.position() / ...` line. It's passing 0 as the vertex count now. Since we use the vertex count to ensure storage in the index buffer, we end up overrunning the buffer on complex monitors if nobody else has grown it out.

This is the source of this type of error that you showed me on Discord:
![2022-06-07_17 09 46](https://user-images.githubusercontent.com/748280/172516107-f0563616-ecef-43f6-b5c4-e67433321e01.png)

Also, this is my fault! I knew we should be including the cursor vertices in the count, and I remember trying to do it this same way. I couldn't figure it out so I left it for somebody else to trip over :P

toad-dev replied:
Well I was wrong again: we have to flip the buffer before passing to `vbo.upload`. I guess we need to calculate the total vertices before the flip. I'd like to clean up some of the other mess I left in this section too:

```patch
--- a/src/main/java/dan200/computercraft/client/render/TileEntityMonitorRenderer.java
+++ b/src/main/java/dan200/computercraft/client/render/TileEntityMonitorRenderer.java
@@ -200,18 +200,17 @@ private static void renderTerminal( @Nonnull MultiBufferSource bufferSource, Pos
                     var buffer = sink.buffer();
 
                     DirectFixedWidthFontRenderer.drawTerminalBackground( sink, 0, 0, terminal, !monitor.isColour(), yMargin, yMargin, xMargin, xMargin );
-                    int backgroundBytes = buffer.position();
+                    monitor.backgroundVertexCount = buffer.position() / vertexSize;
                     DirectFixedWidthFontRenderer.drawTerminalForeground( sink, 0, 0, terminal, !monitor.isColour(), yMargin, yMargin, xMargin, xMargin );
-                    int termVertices = buffer.position() / vertexSize;
-                    monitor.backgroundVertexCount = backgroundBytes / vertexSize;
-                    monitor.foregroundVertexCount = termVertices - monitor.backgroundVertexCount;
+                    monitor.foregroundVertexCount = (buffer.position() / vertexSize) - monitor.backgroundVertexCount;
 
                     // If the cursor is visible, we append it to the end of our buffer. When rendering, we can either
                     // render n or n+1 quads and so toggle the cursor on and off.
                     DirectFixedWidthFontRenderer.drawCursor( sink, 0, 0, terminal, !monitor.isColour() );
 
+                    int totalVertices = buffer.position() / vertexSize;
                     buffer.flip();
-                    vbo.upload( buffer.position() / vertexSize, RenderTypes.MONITOR.mode(), sink.getFormat(), buffer );
+                    vbo.upload( totalVertices, RenderTypes.MONITOR.mode(), sink.getFormat(), buffer );
                 }
 
                 // TODO Figure out how to setup the inverse view rotation matrix properly.
```

toad-dev commented on src/main/java/dan200/computercraft/shared/integration/IrisCompat.java:
```diff
@@ -17,12 +24,60 @@ public boolean isRenderingShadowPass()
+        @Override
+        public DirectFixedWidthFontRenderer.QuadSink getVertexSink( int vertexCount, IntFunction<ByteBuffer> makeBuffer )
+        {
+            return IrisApi.getInstance().getMinorApiRevision() >= 1
+                ? new IrisQuadSink( vertexCount, makeBuffer )
+                : super.getVertexSink( vertexCount, makeBuffer );
```
Yay for compatibility across versions!

toad-dev reviewed (changes requested):
Thank you so much Squid! I've been really busy and wasn't sure when I would be able to work on this. I only saw the one issue. Interested to hear about performance and JITWatch things. I learned quite a bit talking to you about optimization stuff last time.

SquidDev committed:
Clean up vertex counting

SquidDev committed:
Fix rendering glitches when index buffer resizes

When an index buffer grows beyond the 2^16 boundary, the type is
promoted from a SHORT to an INT. However, this type is not updated in
any of the existing DirectVertexBuffers, so we mis-interpret the index
buffer and render garbage.

To fix this, we now store the index buffer object instead (and also
acquire it during upload), and then read the type from that (which is
guaranteed to be always correct).

This is a little more aligned to how 1.19 does things (though 1.19 does
bind during upload rather than first draw).

SquidDev committed:
Remove debugging line

SquidDev commented:
Good enough (TM).

![A screenshot of JITWatch showing quad being inlined](https://user-images.githubusercontent.com/4346137/172700364-726b7e32-4f6d-4579-8278-3278910949de.png)

![A flamechart showing perf is roughly the same, I guess??](https://user-images.githubusercontent.com/4346137/172701254-20f11743-b0f2-4c31-924e-0486eb2c712a.png)

![A CC:R monitor showing Waver, using Iris shaders](https://user-images.githubusercontent.com/4346137/172702661-932ec72f-4d81-4229-86a1-edaad9a7c162.png)

toad-dev commented:
Looks good to me 👍

SquidDev commented:
Ahhh. I targetted the wrong branch for this. Don't know if you want me to recreate on top of `/1.18.2`, or just merge `/stable` into `/1.18.2`?

Merith-TK commented:
gah! yeah just merge stable into 1.18.2
